services/route_manager/src/main/java/com/futurewei/alcor/route/service/Impl/RouterServiceImpl.java

```diff
@@ -0,0 +1,316 @@
+            vpcRouteTables.add(inRoutetable);
+            router.setVpcRouteTable(vpcRouteTables);
+            this.routerDatabaseService.addRouter(router);
+
+            return inRoutetable;
+        } else {
+            List<RouteEntry> routeEntities = routeTable.getRouteEntities();
+            List<RouteEntry> inRouteEntities = inRoutetable.getRouteEntities();
+
+            for (RouteEntry routeEntry : routeEntities) {
+                if (!inRouteEntities.contains(routeEntry)) {
+                    // TODO: check if existing rules are currently used by other subnet’s routing table
+                }
+            }
+            routeTable.setRouteEntities(inRouteEntities);
+            vpcRouteTables.add(routeTable);
+            router.setVpcRouteTable(vpcRouteTables);
+            this.routerDatabaseService.addRouter(router);
+
+            return routeTable;
+        }
+
+    }
+
+    @Override
+    public List<RouteTable> getVpcRouteTables(String projectId, String vpcId) throws CanNotFindVpc {
+        VpcWebJson vpcResponse = this.vpcRouterToVpcService.getVpcWebJson(projectId, vpcId);
+        VpcEntity vpcEntity = vpcResponse.getNetwork();
+        Router router = vpcEntity.getRouter();
+        if (router == null) {
+            return new ArrayList<RouteTable>();
+        }
+        return router.getVpcRouteTable();
+    }
+
+    @Override
+    public RouteTable getOrCreateSubnetRouteTable(String projectId, String subnetId) throws CacheException, ExistMultipleSubnetRouteTable, DatabasePersistenceException {
+        RouteTable routeTable = null;
+
+        Map<String, RouteTable> routeTableMap = null;
+        Map<String, Object[]> queryParams = new HashMap<>();
+        Object[] values = new Object[1];
+        values[0] = subnetId;
+        queryParams.put("owner", values);
+
+        routeTableMap = this.routeTableDatabaseService.getAllRouteTables(queryParams);
+        if (routeTableMap == null) {
+            routeTableMap = new HashMap<>();
+        }
+
+        if (routeTableMap.size() == 0) {
+            // TODO: call subnet API to create Subnet route table and route rule
+            return routeTable;
+        } else if (routeTableMap.size() > 1) {
+            throw new ExistMultipleSubnetRouteTable();
+        } else {
+            for (Map.Entry<String, RouteTable> entry : routeTableMap.entrySet()) {
+                routeTable = (RouteTable)entry.getValue();
+            }
+        }
+
+        return routeTable;
+    }
+
+    @Override
+    public RouteTable updateSubnetRouteTable(String projectId, String subnetId, RouteTableWebJson resource) throws CacheException, DatabasePersistenceException, ExistMultipleSubnetRouteTable {
+        RouteTable routeTable = new RouteTable();
+        RouteTable inRoutetable = resource.getRoutetable();
+        // Get or create a router for a Subnet
+        routeTable = getOrCreateSubnetRouteTable(projectId, subnetId);
```

Comment:
If we follow the logic and comments I gave before, we should able to get the subnet's routing table with subnet_id and update its routing rules. If the subnet doesn't have its own route table, that's mean it currently uses VPC's default route table. Then we just create a new route table for it and push routing rules it has into the route table.